### PR TITLE
esp32s2: Add basic support to SPIFLASH

### PR DESCRIPTION
**Pull request by acassis**
## Summary
Add basic support to SPIFLASH
## Impact
User will be able to use ESP32-S2 SPIFLASH
## Testing
esp32s2-saola-1

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+            .erase  = esp32s2_erase,
+            .bread  = esp32s2_bread,
+            .bwrite = esp32s2_bwrite,
+            .read   = esp32s2_read,
+            .ioctl  = esp32s2_ioctl,
+#ifdef CONFIG_MTD_BYTE_WRITE
+            .write  = esp32s2_write,
+#endif
+            .name   = "esp32s2_spiflash"
+          },
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c:125]: remove extra spaces

File: arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+            .erase  = esp32s2_erase,
+            .bread  = esp32s2_bread_decrypt,
+            .bwrite = esp32s2_bwrite_encrypt,
+            .read   = esp32s2_read_decrypt,
+            .ioctl  = esp32s2_ioctl,
+#ifdef CONFIG_MTD_BYTE_WRITE
+            .write  = NULL,
+#endif
+            .name   = "esp32s2_spiflash_encrypt"
+          },
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c:142]: remove extra spaces

File: arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+ *   nbytes - data number
+ *   buffer - data buffer pointer
+ *
+ * Returned Value:
+ *   Read data bytes if success or a negative value if fail.
+ *
+ ****************************************************************************/
+
+static ssize_t esp32s2_read(struct mtd_dev_s *dev, off_t offset,
+                          size_t nbytes, uint8_t *buffer)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c:235]: add 2 spaces

File: arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+ *   nblocks    - blocks number
+ *   buffer     - data buffer pointer
+ *
+ * Returned Value:
+ *   Read block number if success or a negative value if fail.
+ *
+ ****************************************************************************/
+
+static ssize_t esp32s2_bread(struct mtd_dev_s *dev, off_t startblock,
+                           size_t nblocks, uint8_t *buffer)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c:286]: add 2 spaces

File: arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+ *
+ * Returned Value:
+ *   Read data bytes if success or a negative value if fail.
+ *
+ ****************************************************************************/
+
+static ssize_t esp32s2_read_decrypt(struct mtd_dev_s *dev,
+                                  off_t offset,
+                                  size_t nbytes,
+                                  uint8_t *buffer)
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_spiflash_mtd.c:341]: add 2 spaces

File: boards/xtensa/esp32s2/common/Kconfig [hunk 6]
```diff
@@ -12,6 +12,47 @@ config ESP32S2_MERGE_BINS
...
+	comment "NXFFS not supported with Flash Encryption"
+		depends on ESP32S2_SECURE_FLASH_ENC_ENABLED
+
+	config ESP32S2_SPIFLASH_SPIFFS
+		bool "SPIFFS"
+		select FS_SPIFFS
+		depends on !ESP32S2_SECURE_FLASH_ENC_ENABLED
+
+	comment "SPIFFS not supported with Flash Encryption"
+		depends on ESP32S2_SECURE_FLASH_ENC_ENABLED
```
Comment (pkarashchenko) [inline boards/xtensa/esp32s2/common/Kconfig:48]: Remove extra TABs

File: boards/xtensa/esp32s2/common/src/esp32s2_board_spiflash.c [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <nuttx/kmalloc.h>
+#include <nuttx/spi/spi.h>
+#include <nuttx/mtd/mtd.h>
+#include <nuttx/fs/nxffs.h>
+#ifdef CONFIG_BCH
+#include <nuttx/drivers/drivers.h>
```
Comment (pkarashchenko) [inline boards/xtensa/esp32s2/common/src/esp32s2_board_spiflash.c:42]:
```suggestion
#  include <nuttx/drivers/drivers.h>
```

**Pull request merged**